### ISSUE TITLE
Uploading a Video not working

### ISSUE THREAD

**Issue by TopResults** [google/google-api-php-client#557]
I am using the your scripting to upload videos to YouTube using the offline method by and storing the refresh token - everything seems to be working except for actually uploading the video itself.

The refresh token is auto generating, I have successful access each time I use the scripting. Here is a copy of the scripting - this version as no resembling of the video they are only 3MB videos and I tried both with no luck. When done processing I my status post back from Google is NULL. I am hoping a fresh set of eyes see something I missed or not entered correctly. Thanks

```
$key = file_get_contents('path to token/token.txt');
set_include_path($_SERVER['DOCUMENT_ROOT'] . '/google-api-php-client/');
require_once 'src/Google/autoload.php';
require_once 'src/Google/Client.php';
require_once 'src/Google/Service/YouTube.php';
$application_name = 'Video Uploader';
$client_secret = 'xxx';
$client_id = 'xxx.apps.googleusercontent.com';
$scope = array('https://www.googleapis.com/auth/youtube.upload', 'https://www.googleapis.com/auth/youtube', 'https://www.googleapis.com/auth/youtubepartner');
//grab video details from database videoTitle, videoDescription, videoTags
try {
	$client = new Google_Client();
    $client->setApplicationName($application_name);
    $client->setClientId($client_id);
    $client->setAccessType('offline');
    $client->setAccessToken($key);
    $client->setScopes($scope);
    $client->setClientSecret($client_secret);
	if ($client->getAccessToken()) {
		if ($client->isAccessTokenExpired()) {
            $newToken = json_decode($client->getAccessToken());
            $client->refreshToken($newToken->refresh_token);
			file_put_contents('path to token/token.txt', $client->getAccessToken());
        }
		$youtube = new Google_Service_YouTube($client);
		$snippet = new Google_Service_YouTube_VideoSnippet();
		$snippet->setTitle($videoTitle);
		$snippet->setDescription($videoDescription);
		$snippet->setTags($videoTags);
		$snippet->setCategoryId("22");
		$status = new Google_Service_YouTube_VideoStatus();
		$status->privacyStatus = "public";
    	$video = new Google_Service_YouTube_Video();
	    $video->setSnippet($snippet);
	    $video->setStatus($status);
		$client->setDefer(true);
		$insertRequest = $youtube->videos->insert("status,snippet", $video);
		$media = new Google_Http_MediaFileUpload(
			$client,
			$insertRequest,
			'video/*',
			null,
			false
		);
		$client->setDefer(false);
		var_dump($status);
		if ($status->status['uploadStatus'] == 'uploaded') {
			$videoId = $status['id'];
			//pass to database
		}
	}
}
catch (Google_Service_Exception $e) {
	$error  = "Caught Google service Exception ".$e->getCode(). " message is ".$e->getMessage()."\n";
	$error .= "Stack trace is ".$e->getTraceAsString()."\n";
	//email errors
}
catch (Exception $e) {
	$error  = "Caught Google service Exception ".$e->getCode(). " message is ".$e->getMessage()."\n";
	$error .= "Stack trace is ".$e->getTraceAsString()."\n";
	//email errors
}
```

**Comment by ianbarber** [comment]
I think you're just missing the actual upload triggering - try calling $client->execute($insertRequest) just before you set defer to false.

**Comment by TopResults** [comment]
Will that was close but still not uploading. I inserted your suggestion and it looks to be running as it takes longer to execute. Than when done no uploaded video and no var_dumps on status or media. Which tells me that caused an error. If I remove your suggestion still no upload but I do get var_dump data.

Any one have issues? I am using the latest version of this package.

**Comment by ianbarber** [comment]
Could you edit your original issue with the updated code that fires the execute?

**Comment by TopResults** [comment]
Yes here is what I am using - which includes your execute code. The var_dumps do not even execute when inserting $client->execute($insertRequest); if I remove this all is OK. I will post those in a separate post.

```
$key = file_get_contents($_SERVER['DOCUMENT_ROOT'] . '/google-api-php-client/token.txt');
set_include_path($_SERVER['DOCUMENT_ROOT'] . '/google-api-php-client/');
require_once 'src/Google/autoload.php';
require_once 'src/Google/Client.php';
require_once 'src/Google/Service/YouTube.php';
$application_name = 'Video Uploader';
$client_secret = 'xxxxxxxxxxxxxxxxxxx';
$client_id = 'xxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com';
$scope = array('https://www.googleapis.com/auth/youtube.upload', 'https://www.googleapis.com/auth/youtube', 'https://www.googleapis.com/auth/youtubepartner');
$videoPath = "/i2v/out/".$row["outputPath"];
try {
	$client = new Google_Client();
    $client->setApplicationName($application_name);
    $client->setClientId($client_id);
    $client->setAccessType('offline');
    $client->setAccessToken($key);
    $client->setScopes($scope);
    $client->setClientSecret($client_secret);
	if ($client->getAccessToken()) {
		if ($client->isAccessTokenExpired()) {
            $newToken = json_decode($client->getAccessToken());
            $client->refreshToken($newToken->refresh_token);
			file_put_contents($_SERVER['DOCUMENT_ROOT'] . '/google-api-php-client/token.txt', $client->getAccessToken());
        }
		$youtube = new Google_Service_YouTube($client);
		$snippet = new Google_Service_YouTube_VideoSnippet();
		$snippet->setTitle($videoTitle);
		$snippet->setDescription($videoDescription);
		$snippet->setTags($videoTags);
		$snippet->setCategoryId("22");
		$status = new Google_Service_YouTube_VideoStatus();
		$status->privacyStatus = "public";
    	$video = new Google_Service_YouTube_Video();
	    $video->setSnippet($snippet);
	    $video->setStatus($status);
		$client->setDefer(true);
		$insertRequest = $youtube->videos->insert("status,snippet", $video);
		$media = new Google_Http_MediaFileUpload(
			$client,
			$insertRequest,
			'video/*',
			null,
			false
		);
		$client->execute($insertRequest);
		$client->setDefer(false);
		var_dump($status);
		var_dump($media);
	}
}
```

**Comment by TopResults** [comment]
Here is the results for the var_dump($status); and var_dump($media); which I only get if execute is not inserted. I xxx out my tokens from these posts - they are being process correctly in the originals.

Thanks so much for helping as maybe I am just missing something (crossing fingers)

```
object(Google_Service_YouTube_VideoStatus)#30 (11) {
  ["internal_gapi_mappings":protected]=>
  array(0) {
  }
  ["embeddable"]=>
  NULL
  ["failureReason"]=>
  NULL
  ["license"]=>
  NULL
  ["privacyStatus"]=>
  string(6) "public"
  ["publicStatsViewable"]=>
  NULL
  ["publishAt"]=>
  NULL
  ["rejectionReason"]=>
  NULL
  ["uploadStatus"]=>
  NULL
  ["modelData":protected]=>
  array(0) {
  }
  ["processed":protected]=>
  array(0) {
  }
}
object(Google_Http_MediaFileUpload)#49 (11) {
  ["mimeType":"Google_Http_MediaFileUpload":private]=>
  string(7) "video/*"
  ["data":"Google_Http_MediaFileUpload":private]=>
  NULL
  ["resumable":"Google_Http_MediaFileUpload":private]=>
  bool(false)
  ["chunkSize":"Google_Http_MediaFileUpload":private]=>
  int(262144)
  ["size":"Google_Http_MediaFileUpload":private]=>
  int(0)
  ["resumeUri":"Google_Http_MediaFileUpload":private]=>
  NULL
  ["progress":"Google_Http_MediaFileUpload":private]=>
  int(0)
  ["client":"Google_Http_MediaFileUpload":private]=>
  object(Google_Client)#1 (9) {
    ["auth":"Google_Client":private]=>
    object(Google_Auth_OAuth2)#3 (4) {
      ["assertionCredentials":"Google_Auth_OAuth2":private]=>
      NULL
      ["state":"Google_Auth_OAuth2":private]=>
      NULL
      ["token":"Google_Auth_OAuth2":private]=>
      array(5) {
        ["access_token"]=>
        string(83) "xxxxxxxxxxxxxxxxxxx"
        ["token_type"]=>
        string(6) "Bearer"
        ["expires_in"]=>
        int(3600)
        ["refresh_token"]=>
        string(66) "xxxxxxxxxxxxxxxxxxx"
        ["created"]=>
        int(1432074295)
      }
      ["client":"Google_Auth_OAuth2":private]=>
      *RECURSION*
    }
    ["io":"Google_Client":private]=>
    object(Google_IO_Curl)#7 (2) {
      ["options":"Google_IO_Curl":private]=>
      array(2) {
        [78]=>
        int(100)
        [13]=>
        int(100)
      }
      ["client":protected]=>
      *RECURSION*
    }
    ["cache":"Google_Client":private]=>
    NULL
    ["config":"Google_Client":private]=>
    object(Google_Config)#2 (1) {
      ["configuration":protected]=>
      array(7) {
        ["application_name"]=>
        string(26) "Video Uploader"
        ["auth_class"]=>
        string(18) "Google_Auth_OAuth2"
        ["io_class"]=>
        string(14) "Google_IO_Curl"
        ["cache_class"]=>
        string(17) "Google_Cache_File"
        ["logger_class"]=>
        string(18) "Google_Logger_Null"
        ["base_path"]=>
        string(26) "https://www.googleapis.com"
        ["classes"]=>
        array(9) {
          ["Google_IO_Abstract"]=>
          array(1) {
            ["request_timeout_seconds"]=>
            int(100)
          }
          ["Google_Logger_Abstract"]=>
          array(4) {
            ["level"]=>
            string(5) "debug"
            ["log_format"]=>
            string(42) "[%datetime%] %level%: %message% %context%
"
            ["date_format"]=>
            string(13) "d/M/Y:H:i:s O"
            ["allow_newlines"]=>
            bool(true)
          }
          ["Google_Logger_File"]=>
          array(3) {
            ["file"]=>
            string(12) "php://stdout"
            ["mode"]=>
            int(416)
            ["lock"]=>
            bool(false)
          }
          ["Google_Http_Request"]=>
          array(2) {
            ["disable_gzip"]=>
            bool(true)
            ["enable_gzip_for_uploads"]=>
            bool(false)
          }
          ["Google_Auth_OAuth2"]=>
          array(13) {
            ["client_id"]=>
            string(72) "xxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com"
            ["client_secret"]=>
            string(24) "xxxxxxxxxxxxxxxxxxx"
            ["redirect_uri"]=>
            string(0) ""
            ["developer_key"]=>
            string(0) ""
            ["hd"]=>
            string(0) ""
            ["prompt"]=>
            string(0) ""
            ["openid.realm"]=>
            string(0) ""
            ["include_granted_scopes"]=>
            string(0) ""
            ["login_hint"]=>
            string(0) ""
            ["request_visible_actions"]=>
            string(0) ""
            ["access_type"]=>
            string(7) "offline"
            ["approval_prompt"]=>
            string(4) "auto"
            ["federated_signon_certs_url"]=>
            string(42) "https://www.googleapis.com/oauth2/v1/certs"
          }
          ["Google_Task_Runner"]=>
          array(5) {
            ["initial_delay"]=>
            int(1)
            ["max_delay"]=>
            int(60)
            ["factor"]=>
            int(2)
            ["jitter"]=>
            float(0.5)
            ["retries"]=>
            int(0)
          }
          ["Google_Service_Exception"]=>
          array(1) {
            ["retry_map"]=>
            array(4) {
              [500]=>
              int(-1)
              [503]=>
              int(-1)
              ["rateLimitExceeded"]=>
              int(-1)
              ["userRateLimitExceeded"]=>
              int(-1)
            }
          }
          ["Google_IO_Exception"]=>
          array(1) {
            ["retry_map"]=>
            array(5) {
              [6]=>
              int(-1)
              [7]=>
              int(-1)
              [28]=>
              int(-1)
              [35]=>
              int(-1)
              [52]=>
              int(-1)
            }
          }
          ["Google_Cache_File"]=>
          array(1) {
            ["directory"]=>
            string(18) "/tmp/Google_Client"
          }
        }
      }
    }
    ["logger":"Google_Client":private]=>
    object(Google_Logger_Null)#5 (4) {
      ["level":protected]=>
      int(100)
      ["logFormat":protected]=>
      string(42) "[%datetime%] %level%: %message% %context%
"
      ["dateFormat":protected]=>
      string(13) "d/M/Y:H:i:s O"
      ["allowNewLines":protected]=>
      bool(true)
    }
    ["deferExecution":"Google_Client":private]=>
    bool(false)
    ["requestedScopes":protected]=>
    array(3) {
      [0]=>
      string(46) "https://www.googleapis.com/auth/youtube.upload"
      [1]=>
      string(39) "https://www.googleapis.com/auth/youtube"
      [2]=>
      string(46) "https://www.googleapis.com/auth/youtubepartner"
    }
    ["services":protected]=>
    array(0) {
    }
    ["authenticated":"Google_Client":private]=>
    bool(false)
  }
  ["request":"Google_Http_MediaFileUpload":private]=>
  object(Google_Http_Request)#32 (15) {
    ["batchHeaders":"Google_Http_Request":private]=>
    array(3) {
      ["Content-Type"]=>
      string(16) "application/http"
      ["Content-Transfer-Encoding"]=>
      string(6) "binary"
      ["MIME-Version"]=>
      string(3) "1.0"
    }
    ["queryParams":protected]=>
    array(2) {
      ["part"]=>
      string(14) "status,snippet"
      ["uploadType"]=>
      string(9) "multipart"
    }
    ["requestMethod":protected]=>
    string(4) "POST"
    ["requestHeaders":protected]=>
    array(2) {
      ["content-type"]=>
      string(36) "multipart/related; boundary=92065581"
      ["authorization"]=>
      string(90) "Bearer xxxxxxxxxxxxxxxxxxx"
    }
    ["baseComponent":protected]=>
    string(33) "https://www.googleapis.com/upload"
    ["path":protected]=>
    string(18) "/youtube/v3/videos"
    ["postBody":protected]=>
    string(808) "--92065581
Content-Type: application/json; charset=UTF-8

{"snippet":{"categoryId":"22","description":"Video description goes here","tags":["Tag One","Tag Two","Tag Three"],"title":"Video title goes here"},"status":{"privacyStatus":"public"}}
--92065581
Content-Type: video/*
Content-Transfer-Encoding: base64


--92065581--"
    ["userAgent":protected]=>
    NULL
    ["canGzip":protected]=>
    NULL
    ["responseHttpCode":protected]=>
    NULL
    ["responseHeaders":protected]=>
    NULL
    ["responseBody":protected]=>
    NULL
    ["expectedClass":protected]=>
    string(28) "Google_Service_YouTube_Video"
    ["expectedRaw":protected]=>
    bool(false)
    ["accessKey"]=>
    NULL
  }
  ["boundary":"Google_Http_MediaFileUpload":private]=>
  bool(false)
  ["httpResultCode":"Google_Http_MediaFileUpload":private]=>
  NULL
}
```

**Comment by TopResults** [comment]
Sorry about that; I just realize what you meant. There is a new coding to download. I been away and just read that email letting me know about your new version. Let me download and try this. BRB

**Comment by TopResults** [comment]
I updated to the latest version and still the same results with no vars being dumped. If I block out that one section, I get the results, just not the ones I was hoping for.

**Comment by TopResults** [comment]
This seems to be where I am having my issue with this script.

```
$status = $media->nextChunk($chunk);
```

If I comment this out the script works but doesn't upload. It use it the script stalls out and there are no errors what so ever, I have PHP setup to track full errors and it isn't detecting nothing.

**Comment by ianbarber** [comment]
Can you see if you can get anywhere with the Drive file upload example: in https://github.com/google/google-api-php-client/blob/master/examples/fileupload.php - should be relatively straightforward to adapt it to a video object instead?

**Comment by TopResults** [comment]
Ok, this looks promising. I am now getting the following error from Google. What I did was turn off throw exceptions within your scripting.

After doing this I get my status results from Google and here is why it isn't uploading.
```
["responseBody":protected]=>
	string(259) "{
 		"error": {
  			"errors": [
   			{
			"domain": "youtube.video",
    			"reason": "invalidDescription",
    			"message": "Bad Request",
    			"locationType": "other",
    			"location": "body.snippet.description"
   			}
  			],
  		"code": 400,
		"message": "Bad Request"
	}
}
```
I have HTML in my description - I have <br /> we not allowed these?

**Comment by ianbarber** [comment]
Aha! I am not familiar with the specifics of  the youtube API, but it a quick check of the API docs: https://developers.google.com/youtube/v3/docs/videos#resource

"The video's description. The property value has a maximum length of 5000 bytes and may contain all valid UTF-8 characters except < and >."

**Comment by TopResults** [comment]
Solved, I was using <br /> which was allowed in JAVA Applets. I switched to PHP and kept the same format. Not allowed.

As for PHP Exceptions - that is just odd that the server I am using has an issue with them and why I wasn't getting any errors to track. I had to go old school and track line by line until I found what gave.

I am using PHP 5.3.3 - either or, all is good and I am uploading. Thanks so much for your patience.
